ext/intl/common/common_date.cpp

```diff
@@ -137,7 +139,8 @@ U_CFUNC int intl_datetime_decompose(zval *z, double *millis, TimeZone **tz,
 			return FAILURE;
 		}
 
-		*millis = U_MILLIS_PER_SECOND * (double)Z_LVAL(retval);
+		datetime = Z_PHPDATE_P(z);
+		*millis = U_MILLIS_PER_SECOND * ((double)Z_LVAL(retval) + datetime->time->f);
```

Comment:
@nikic @derickr I was thinking about this too, but at the same time I took this idea from here https://github.com/andrewnester/php-src/blob/1bf8cfbb9a136ee64303206446ec7e2e97f7ec4c/ext/intl/common/common_date.cpp#L150